### ISSUE TITLE
openCL error

### ISSUE THREAD

**Issue by adzweig** [lightvector/KataGo#370]
I have tried to get the OpenCL version working with Lizzie as a front end. I obtain two error messages: 
1) Beignet: "unable to find good values for local_work_size[i], please provide\n" " local_work_size[] explicitly, you can find good values with\n" " trial-and-error method."
2) terminate called after throwing an instance of 'StringError' 
  what():  OpenCL error at /home/adrian/Go/bin/KataGo/cpp/neuralnet/openclbackend.cpp, func err, line 1188, error CL_MEM_OBJECT_ALLOCATION_FAILURE 

The CUDA implementations work fine, but I would like to use the OpenGL version, in order to make use of the second GPU to speed up performance. 

I obtain the same OpenCL error when working with a downloaded executable or when I use one built on my platform ( Ubuntu 20.04) 
Any ideas how I should proceed?

thanks a lot

**Comment by adzweig** [comment]
./katago version returns: 
----------------
adrian@127:~/Go/bin/LeelaZero/KataGoOpenGL$ ./katago version
KataGo v1.7.0
Git revision: 684de9a5af164d875c0abd373aee7539b488718a
Compile Time: Dec 20 2020 18:59:12
Using OpenCL backend
-------------------
./katago gtp returns: 
----------------------
adrian@127:~/Go/bin/LeelaZero/KataGoOpenGL$ ./katago gtp
KataGo v1.7.0
Using TrompTaylor rules initially, unless GTP/GUI overrides this
terminate called after throwing an instance of 'std::length_error'
  what():  basic_string::_M_replace_aux
Aborted (core dumped)
----------------------
Any ideas what could be the origin of this run time error?

**Comment by adzweig** [comment]
I managed to get the openGL version working partially: 
> when uncommenting and setting the parameter "nnmaxBatchsize" katago seems to work with "Katrain" as GUI, but not with "Lizzie" as GUI
> when commenting away and thus not setting the parameter "nnmaxBatchsize" katago seems to work with "Lizzie" as GUI, but not with "Katrain" as GUI
However running from the command line I obtain the same error with both config files:
-----------
adrian@127:~/Go/bin/LeelaZero/KataGoOpenGL$ ./katago benchmark -config analysis_config.cfg
2020-12-21 13:03:39+0100: Loading model and initializing benchmark...
WARNING: Your nnMaxBatchSize is hardcoded to 384, ignoring it and assuming it is >= threads, for this benchmark.
2020-12-21 13:03:39+0100: nnRandSeed0 = 17908438132967980509
2020-12-21 13:03:39+0100: After dedups: nnModelFile0 = /home/adrian/Go/bin/LeelaZero/KataGoOpenGL/default_model.bin.gz useFP16 auto useNHWC auto
terminate called after throwing an instance of 'std::length_error'
  what():  basic_string::_M_replace_aux
Aborted (core dumped)
adrian@127:~/Go/bin/LeelaZero/KataGoOpenGL$ ./katago benchmark -config katago_gtp10.cfg
2020-12-21 13:03:57+0100: Loading model and initializing benchmark...
2020-12-21 13:03:57+0100: nnRandSeed0 = 14153785799760072644
2020-12-21 13:03:57+0100: After dedups: nnModelFile0 = /home/adrian/Go/bin/LeelaZero/KataGoOpenGL/default_model.bin.gz useFP16 auto useNHWC auto
terminate called after throwing an instance of 'std::length_error'
  what():  basic_string::_M_replace_aux
Aborted (core dumped)
----------

**Comment by lightvector** [comment]
Out of curiosity, does Leela Zero work on your GPU using OpenCL? If it doesn't, that's definitely a sign that your GPU drivers aren't happy and maybe don't support OpenCL - what is your GPU? If it does, then it is slightly less likely to be that, although it still could be, since KataGo does a few kinds of OpenCL operations on the GPU that Leela Zero might not do. If this is the situation, the only fix is maybe to upgrade your drivers or get a newer GPU.

Anyways there is a chance that your OpenCL implementation is failing but well-behaved, but returning a particular kind of error code or situation that KataGo is failing to handle properly. For example, maybe it is returning 0 devices, or 0 platforms, or something, and maybe KataGo does some invalid memory stuff when that happens. I'm not sure.

Secondly, there is a possibility is that your OpenCL implementation *is* working instead of failing, but unlike OpenCL on other more recent GPUs, is doing something that KataGo isn't handling properly.

I'd recommend trying Leela Zero first to see if it's a fundamental OpenCL thing. If it seems to be KataGo+OpenCL specific, then I think the next step, if you're willing, and given that you seem to be experienced enough to build from source, is to run KataGo within gdb as well. Run within gdb, let it crash, and then give me the output of `backtrace` and maybe I can figure out what's going on. Or, if you're feeling adventurous, you can poke around in the code yourself on different stack frames and try to work it out too, since you have the advantage of being able to reproduce the crash and work interactively (and/or to also add debug print statements to the code around the functions that the backtrace showed had crashed).

If there is a bug in KataGo to be fixed here (even if it's just more friendly error-reporting of an underlying problem in OpenCL), I'm happy to fix it.

**Comment by adzweig** [comment]
thanks a lot, will proceed as suggested. Leelazero works fine. I rebuilt it and ran ./tests, cf below.
------------------
adrian@127:~/Go/bin/LeelaZero/leela-zero/build$ ./tests
Running main() from gtest_main.cc
[==========] Running 13 tests from 2 test cases.
[----------] Global test environment set-up.
BLAS Core: built-in Eigen 3.3.7 library.
Detecting residual layers...v1...8 channels...1 blocks.
Initializing OpenCL (autodetecting precision).
Detected 2 OpenCL platforms.
Platform version: OpenCL 2.0 beignet 1.3
Platform profile: FULL_PROFILE
Platform name:    Intel Gen OCL Driver
Platform vendor:  Intel
Device ID:     0
Device name:   Intel(R) HD Graphics IvyBridge GT1
Device type:   GPU
Device vendor: Intel
Device driver: 1.3
Device speed:  1000 MHz
Device cores:  6 CU
Device score:  620
Platform version: OpenCL 1.2 CUDA 11.0.228
Platform profile: FULL_PROFILE
Platform name:    NVIDIA CUDA
Platform vendor:  NVIDIA Corporation
Device ID:     1
Device name:   GeForce GTX 1050 Ti
Device type:   GPU
Device vendor: NVIDIA Corporation
Device driver: 450.80.02
Device speed:  1455 MHz
Device cores:  6 CU
Device score:  1112
Selected platform: NVIDIA CUDA
Selected device: GeForce GTX 1050 Ti
with OpenCL 1.2 capability.
Half precision compute support: No.
Tensor Core support: No.
Detected 2 OpenCL platforms.
Platform version: OpenCL 2.0 beignet 1.3
Platform profile: FULL_PROFILE
Platform name:    Intel Gen OCL Driver
Platform vendor:  Intel
Device ID:     0
Device name:   Intel(R) HD Graphics IvyBridge GT1
Device type:   GPU
Device vendor: Intel
Device driver: 1.3
Device speed:  1000 MHz
Device cores:  6 CU
Device score:  620
Platform version: OpenCL 1.2 CUDA 11.0.228
Platform profile: FULL_PROFILE
Platform name:    NVIDIA CUDA
Platform vendor:  NVIDIA Corporation
Device ID:     1
Device name:   GeForce GTX 1050 Ti
Device type:   GPU
Device vendor: NVIDIA Corporation
Device driver: 450.80.02
Device speed:  1455 MHz
Device cores:  6 CU
Device score:  1112
Selected platform: NVIDIA CUDA
Selected device: GeForce GTX 1050 Ti
with OpenCL 1.2 capability.
Half precision compute support: No.
Tensor Core support: No.
Loaded existing SGEMM tuning.
Wavefront/Warp size: 32
Max workgroup size: 1024
Max workgroup dimensions: 1024 1024 64 
Loaded existing SGEMM tuning.
Wavefront/Warp size: 32
Max workgroup size: 1024
Max workgroup dimensions: 1024 1024 64 
Using OpenCL single precision (less than 5% slower than half).
Detected 2 OpenCL platforms.
Platform version: OpenCL 2.0 beignet 1.3
Platform profile: FULL_PROFILE
Platform name:    Intel Gen OCL Driver
Platform vendor:  Intel
Device ID:     0
Device name:   Intel(R) HD Graphics IvyBridge GT1
Device type:   GPU
Device vendor: Intel
Device driver: 1.3
Device speed:  1000 MHz
Device cores:  6 CU
Device score:  620
Platform version: OpenCL 1.2 CUDA 11.0.228
Platform profile: FULL_PROFILE
Platform name:    NVIDIA CUDA
Platform vendor:  NVIDIA Corporation
Device ID:     1
Device name:   GeForce GTX 1050 Ti
Device type:   GPU
Device vendor: NVIDIA Corporation
Device driver: 450.80.02
Device speed:  1455 MHz
Device cores:  6 CU
Device score:  1112
Selected platform: NVIDIA CUDA
Selected device: GeForce GTX 1050 Ti
with OpenCL 1.2 capability.
Half precision compute support: No.
Tensor Core support: No.
Loaded existing SGEMM tuning.
Wavefront/Warp size: 32
Max workgroup size: 1024
Max workgroup dimensions: 1024 1024 64 
Setting max tree size to 4385 MiB and cache size to 487 MiB.
[----------] 4 tests from UtilsTest
[ RUN      ] UtilsTest.CeilMultiple
[       OK ] UtilsTest.CeilMultiple (0 ms)
[ RUN      ] UtilsTest.RandFix
[       OK ] UtilsTest.RandFix (0 ms)
[ RUN      ] UtilsTest.Randuint64_lastEightBits
[       OK ] UtilsTest.Randuint64_lastEightBits (0 ms)
[ RUN      ] UtilsTest.Randuint64_max
[       OK ] UtilsTest.Randuint64_max (0 ms)
[----------] 4 tests from UtilsTest (0 ms total)

[----------] 9 tests from LeelaTest
[ RUN      ] LeelaTest.Startup
[       OK ] LeelaTest.Startup (0 ms)
[ RUN      ] LeelaTest.DefaultHash
[       OK ] LeelaTest.DefaultHash (0 ms)
[ RUN      ] LeelaTest.Transposition
[       OK ] LeelaTest.Transposition (0 ms)
[ RUN      ] LeelaTest.KoPntNotSame
[       OK ] LeelaTest.KoPntNotSame (1 ms)
[ RUN      ] LeelaTest.MoveOnOccupiedPnt
[       OK ] LeelaTest.MoveOnOccupiedPnt (0 ms)
[ RUN      ] LeelaTest.TimeControl
[       OK ] LeelaTest.TimeControl (11 ms)
[ RUN      ] LeelaTest.TimeControl2
[       OK ] LeelaTest.TimeControl2 (6 ms)
[ RUN      ] LeelaTest.AnalyzeParse
[       OK ] LeelaTest.AnalyzeParse (0 ms)
[ RUN      ] LeelaTest.AnalyzeParseMinmoves
[       OK ] LeelaTest.AnalyzeParseMinmoves (2 ms)
[----------] 9 tests from LeelaTest (20 ms total)

[----------] Global test environment tear-down
[==========] 13 tests from 2 test cases ran. (6060 ms total)
[  PASSED  ] 13 tests.

**Comment by adzweig** [comment]
building katago gives a few warnings: 
adrian@127:~/Go/bin/LeelaZero/KataGo/KataGo/cpp$ cmake . -DUSE_BACKEND=OPENCL 
-- The C compiler identification is GNU 7.5.0
-- The CXX compiler identification is GNU 9.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Building 'katago' executable for GTP engine and other tools.
-- -DUSE_BACKEND=OPENCL, using OpenCL backend.
-- Including Git revision in the compiled executable, specify -DNO_GIT_REVISION=1 to disable
-- Found Git: /usr/bin/git (found version "2.25.1") 
-- Looking for CL_VERSION_2_2
-- Looking for CL_VERSION_2_2 - found
-- Found OpenCL: /usr/lib/x86_64-linux-gnu/libOpenCL.so (found version "2.2") 
-- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.11") 
CMake Warning at CMakeLists.txt:363 (message):
  WARNING: libzip library was NOT found.  KataGo should still work
  for GTP/matches/analysis if everything else is good, but selfplay for
  writing training data will not be possible.


-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Configuring done
-- Generating done
-- Build files have been written to: /home/adrian/Go/bin/LeelaZero/KataGo/KataGo/cpp
---------------
now proceeding to gdb....

**Comment by adzweig** [comment]
I did some further exploration: The issue appears to be the size of nnmaxBatchSize. If it is too large, the engine dies under Lizzie. Making it smaller or leaving it undefined eliminates the problem. Katrain seems to not allow too large values of this parameter, but asks you to specify it.
I now managed to get both the CUDA and the OpenGL versions to work, the latter using both GPUs in my system ( :-) )
